From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manoj Sekar <quic_sekar@quicinc.com>
Date: Mon, 26 Feb 2024 18:21:33 +0530
Subject: [PATCH 859/860] Multi-AP: Add hostapd config option to disallow
 certain profiles
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add a new config option "multi_ap_client_disallow" to control allowing
backhaul STA with certain profiles alone to associate. This is done to
adhere to Wi-Fi EasyMesh specification which defined rules to
allow/disallow association of backhaul STA of certain profiles.

Upstream-Status: Backport [210c2b4bd75e59e2956635f3800839b947ce5616]
Signed-off-by: Manoj Sekar <quic_sekar@quicinc.com>
Signed-off-by: Petr Štetiar <petr.stetiar@prplfoundation.org>
---
 hostapd/config_file.c        | 10 ++++++++++
 hostapd/hostapd.conf         |  7 +++++++
 src/ap/ap_config.h           |  5 +++++
 src/ap/ieee802_11.c          | 29 +++++++++++++++++++++++++++++
 src/common/ieee802_11_defs.h |  2 ++
 5 files changed, 53 insertions(+)

--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -4601,6 +4601,16 @@ static int hostapd_config_fill(struct ho
 			return -1;
 		}
 		bss->multi_ap_profile = val;
+	} else if (os_strcmp(buf, "multi_ap_client_disallow") == 0) {
+		int val = atoi(pos);
+
+		if (val < 0 || val > 3) {
+			wpa_printf(MSG_ERROR,
+				   "Line %d: Invalid multi_ap_client_allow '%s'",
+				   line, buf);
+			return -1;
+		}
+		bss->multi_ap_client_disallow = val;
 	} else if (os_strcmp(buf, "multi_ap_vlanid") == 0) {
 		int val = atoi(pos);
 
--- a/hostapd/hostapd.conf
+++ b/hostapd/hostapd.conf
@@ -2512,6 +2512,13 @@ own_ip_addr=127.0.0.1
 # A valid non-zero VLAN ID will be used to update Default IEEE 802.1Q Setting
 #multi_ap_vlanid=0
 
+# Multi-AP client disallow
+# Used to disallow profile specific backhaul STA association
+# Bitmap of the disallowed Profile-X profiles
+# 1 = Profile-1 Backhaul STA association disallowed
+# 2 = Profile-2 Backhaul STA association disallowed
+#multi_ap_client_disallow=0
+
 # WPS UPnP interface
 # If set, support for external Registrars is enabled.
 #upnp_iface=br0
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -795,6 +795,11 @@ struct hostapd_bss_config {
 #define FRONTHAUL_BSS 2
 	int multi_ap; /* bitmap of BACKHAUL_BSS, FRONTHAUL_BSS */
 	int multi_ap_profile;
+	/* Multi-AP Profile-1 clients not allowed to connect */
+#define PROFILE1_CLIENT_ASSOC_DISALLOW BIT(0)
+	/* Multi-AP Profile-2 clients not allowed to connect */
+#define PROFILE2_CLIENT_ASSOC_DISALLOW BIT(1)
+	unsigned int multi_ap_client_disallow;
 
     /* Primary VLAN ID to use in Multi-AP */
     int multi_ap_vlanid;
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -99,6 +99,15 @@ static u8 * hostapd_eid_multi_ap(struct
 	if (hapd->conf->multi_ap & FRONTHAUL_BSS)
 		multi_ap.capability |= MULTI_AP_FRONTHAUL_BSS;
 
+       if (hapd->conf->multi_ap_client_disallow &
+           PROFILE1_CLIENT_ASSOC_DISALLOW)
+               multi_ap.capability |=
+                       MULTI_AP_PROFILE1_BACKHAUL_STA_DISALLOWED;
+       if (hapd->conf->multi_ap_client_disallow &
+           PROFILE2_CLIENT_ASSOC_DISALLOW)
+               multi_ap.capability |=
+                       MULTI_AP_PROFILE2_BACKHAUL_STA_DISALLOWED;
+
     multi_ap.profile = hapd->conf->multi_ap_profile;
     multi_ap.vlanid = hapd->conf->multi_ap_vlanid;
     
@@ -3443,6 +3452,26 @@ static u16 check_multi_ap(struct hostapd
 			       "Multi-AP IE with unexpected value 0x%02x",
 			       multi_ap.capability);
 
+	if (multi_ap.profile == MULTI_AP_PROFILE_1 &&
+	    (hapd->conf->multi_ap_client_disallow &
+	     PROFILE1_CLIENT_ASSOC_DISALLOW)) {
+		hostapd_logger(hapd, sta->addr,
+			       HOSTAPD_MODULE_IEEE80211,
+			       HOSTAPD_LEVEL_INFO,
+			       "Multi-AP Profile-1 clients not allowed");
+		return WLAN_STATUS_ASSOC_DENIED_UNSPEC;
+	}
+
+	if (multi_ap.profile >= MULTI_AP_PROFILE_2 &&
+	    (hapd->conf->multi_ap_client_disallow &
+	     PROFILE2_CLIENT_ASSOC_DISALLOW)) {
+		hostapd_logger(hapd, sta->addr,
+			       HOSTAPD_MODULE_IEEE80211,
+			       HOSTAPD_LEVEL_INFO,
+			       "Multi-AP Profile-2 clients not allowed");
+		return WLAN_STATUS_ASSOC_DENIED_UNSPEC;
+	}
+
 	if (!(multi_ap.capability & MULTI_AP_BACKHAUL_STA)) {
 		if (hapd->conf->multi_ap & FRONTHAUL_BSS)
 			return WLAN_STATUS_SUCCESS;
--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -1396,6 +1396,8 @@ struct ieee80211_ampe_ie {
 #define MULTI_AP_PROFILE_SUB_ELEM_TYPE 0x07
 #define MULTI_AP_VLAN_SUB_ELEM_TYPE 0x08
 
+#define MULTI_AP_PROFILE2_BACKHAUL_STA_DISALLOWED BIT(2)
+#define MULTI_AP_PROFILE1_BACKHAUL_STA_DISALLOWED BIT(3)
 #define MULTI_AP_TEAR_DOWN BIT(4)
 #define MULTI_AP_FRONTHAUL_BSS BIT(5)
 #define MULTI_AP_BACKHAUL_BSS BIT(6)
